Add support of custom headers to OBDCommand

Ircama opened the pull request:
**Add support of custom headers**

Allow the optional `header` argument to [`OBDCommand`](https://github.com/brendan-w/python-OBD/blob/master/docs/Custom%20Commands.md), which tells python-OBD to use a custom header when querying ([`query`](https://github.com/brendan-w/python-OBD/blob/master/docs/Connections.md#querycommand-forcefalse)) a command with this `OBDCommand` argument set to a value different from '7E0'.

If not set, python-OBD assumes that the default 7E0 header is needed for querying the command.

Specifically, the code assumes that there is a default 7E0 header (`ECU_HEADER.ENGINE`) which refers to ECU#1. So, a preliminary `AT SH 7E0` is not issued when querying basic commands like `SPEED`.
 
The switch between default and custom header (and vice versa) is automatically done by python-OBD.

This code is different from https://github.com/brendan-w/python-OBD/pull/65, where the header is controlled at a connection level and not per-command.

alistair23 commented:
What is the benefit of allowing a custom header? Can you update the documentation with information about when and how people should use it?

Ircama commented:
Here is a sample program testing headers with a Toyota Hybrid car, which supports custom commands to query the tank levels and showing that commands with different headers can be mixed:

```
import time
import obd
import sys

from obd import OBDCommand, Unit
from obd.protocols import ECU
from obd.utils import bytes_to_int

ICE_ECU = b'7C0'

main_fuel = OBDCommand(
    "MAIN_TANK",
    "Fuel Level - Main tank",
    b"2129",
    3,
    lambda m: (m[0].data[2] / 4.0) * Unit.liter,
    ECU.ALL,
    True,
    header = ICE_ECU)

sub_fuel = OBDCommand(
    "SUB_TANK",
    "Fuel Level - Sub tank",
    b"212A",
    3,
    lambda m: (m[0].data[2] / 4.0) * Unit.liter,
    ECU.ALL,
    True,
    header = ICE_ECU)

obd.logger.setLevel(obd.logging.DEBUG)

connection = obd.Async(sys.argv[1])
connection.watch(obd.commands.SPEED, force=True)
connection.watch(obd.commands.RPM, force=True)
connection.watch(main_fuel, force=True)
connection.watch(sub_fuel, force=True)
connection.start()

while True:
    print(connection.query(obd.commands.RPM))
    print("")
    print(connection.query(obd.commands.SPEED))
    print("")
    print(connection.query(main_fuel))
    print("")
    print(connection.query(sub_fuel))
    print("\n---------------------------------\n")
    time.sleep(2)
```

Notice that `OBDCommand` defines `header = ICE_ECU`.

The output of the program without traces is the following:

````
2900.0 revolutions_per_minute

16 kph

5.25 liter

4.5 liter

---------------------------------
````

The traces are:

```
[obd.obd] ======================= python-OBD (v0.7.0) =======================
[obd.obd] Explicit port defined
[obd.elm327] Initializing ELM327: PORT=/dev/pts/1 BAUD=auto PROTOCOL=auto
[obd.elm327] Detected pseudo terminal, skipping baudrate setup
[obd.elm327] write: b'ATZ\r'
[obd.elm327] wait: 1 seconds
[obd.elm327] read: b'\r\rELM327 v1.5\r\r>'
[obd.elm327] write: b'ATE0\r'
[obd.elm327] read: b'OK\r\r>'
[obd.elm327] write: b'ATH1\r'
[obd.elm327] read: b'OK\r\r>'
[obd.elm327] write: b'ATL0\r'
[obd.elm327] read: b'OK\r\r>'
[obd.elm327] write: b'ATSP0\r'
[obd.elm327] read: b'OK\r\r>'
[obd.elm327] write: b'0100\r'
[obd.elm327] read: b'SEARCHING...\r7EA 06 41 00 98 3A 80 13 \r7E8 06 41 00 BE 3F A8 13 \r\r>'
[obd.elm327] write: b'ATDPN\r'
[obd.elm327] read: b'A6\r\r>'
[obd.protocols.protocol] map ECU 0 --> ENGINE
[obd.protocols.protocol] map ECU 2 --> UNKNOWN
[obd.elm327] Connected Successfully: PORT=/dev/pts/1 BAUD=9600 PROTOCOL=6
[obd.obd] querying for supported commands
[obd.obd] Sending command: b'0100': Supported PIDs [01-20]
[obd.elm327] write: b'0100\r'
[obd.elm327] read: b'7EA 06 41 00 98 3A 80 13 \r7E8 06 41 00 BE 3F A8 13 \r\r>'
[obd.obd] Sending command: b'0120': Supported PIDs [21-40]
[obd.elm327] write: b'0120\r'
[obd.elm327] read: b'7EA 06 41 20 80 01 A0 01 \r7E8 06 41 20 90 15 B0 15 \r\r>'
[obd.obd] Sending command: b'0140': Supported PIDs [41-60]
[obd.elm327] write: b'0140\r'
[obd.elm327] read: b'7EA 06 41 40 44 CC 00 21 \r7E8 06 41 40 7A 1C 80 00 \r\r>'
[obd.obd] Sending command: b'0600': Supported MIDs [01-20]
[obd.elm327] write: b'0600\r'
[obd.elm327] read: b'7E8 06 46 00 C0 00 00 01 \r\r>'
[obd.obd] Sending command: b'0620': Supported MIDs [21-40]
[obd.elm327] write: b'0620\r'
[obd.elm327] read: b'7E8 06 46 20 80 00 80 01 \r\r>'
[obd.obd] Sending command: b'0640': Supported MIDs [41-60]
[obd.elm327] write: b'0640\r'
[obd.elm327] read: b'7E8 06 46 40 00 00 00 01 \r\r>'
[obd.obd] Sending command: b'0660': Supported MIDs [61-80]
[obd.elm327] write: b'0660\r'
[obd.elm327] read: b'7E8 06 46 60 00 00 00 01 \r\r>'
[obd.obd] Sending command: b'0680': Supported MIDs [81-A0]
[obd.elm327] write: b'0680\r'
[obd.elm327] read: b'7E8 06 46 80 00 00 00 01 \r\r>'
[obd.obd] Sending command: b'06A0': Supported MIDs [A1-C0]
[obd.elm327] write: b'06A0\r'
[obd.elm327] read: b'7E8 06 46 A0 F8 00 00 00 \r\r>'
[obd.obd] finished querying with 97 commands supported
[obd.obd] ===================================================================
[obd.asynchronous] Watching command: b'010D': Vehicle Speed
[obd.asynchronous] Watching command: b'010C': Engine RPM
[obd.asynchronous] Watching command: b'2129': Fuel Level - Main tank
[obd.asynchronous] Watching command: b'212A': Fuel Level - Sub tank
[obd.asynchronous] Starting async thread
[obd.obd] Sending command: b'010D': Vehicle Speed
[obd.elm327] write: b'010D\r'
[obd.elm327] read: b'7E8 03 41 0D 04 \r7EA 03 41 0D 04 \r\r>'
[obd.obd] Sending command: b'010C': Engine RPM
[obd.elm327] write: b'010C\r'
[obd.elm327] read: b'7E8 04 41 0C 09 B0 \r7EA 04 41 0C 09 B0 \r\r>'
[obd.elm327] write: b'AT SH 7C0 \r'
[obd.elm327] read: b'OK\r\r>'
[obd.obd] Sending command: b'2129': Fuel Level - Main tank
[obd.elm327] write: b'2129\r'
[obd.elm327] read: b'7C8 03 61 29 15 \r\r>'
[obd.obd] Sending command: b'212A': Fuel Level - Sub tank
[obd.elm327] write: b'212A\r'
[obd.elm327] read: b'7C8 03 7F 21 12 \r\r>'
[obd.elm327] write: b'AT SH 7E0 \r'
[obd.elm327] read: b'OK\r\r>'
[obd.obd] Sending command: b'010D': Vehicle Speed
[obd.elm327] write: b'010D2\r'
[obd.elm327] read: b'7E8 03 41 0D 08 \r7EA 03 41 0D 08 \r\r>'
[obd.obd] Sending command: b'010C': Engine RPM
[obd.elm327] write: b'010C2\r'
[obd.elm327] read: b'7E8 04 41 0C 0A 00 \r7EA 04 41 0C 0A 00 \r\r>'
[obd.elm327] write: b'AT SH 7C0 \r'
[obd.elm327] read: b'OK\r\r>'
[obd.obd] Sending command: b'2129': Fuel Level - Main tank
[obd.elm327] write: b'21291\r'
[obd.elm327] read: b'7C8 03 61 29 15 \r\r>'
[obd.obd] Sending command: b'212A': Fuel Level - Sub tank
[obd.elm327] write: b'212A1\r'
[obd.elm327] read: b'7C8 03 7F 21 12 \r\r>'
[obd.elm327] write: b'AT SH 7E0 \r'
[obd.elm327] read: b'OK\r\r>'
[obd.obd] Sending command: b'010D': Vehicle Speed
[obd.elm327] write: b'010D2\r'
[obd.elm327] read: b'7E8 03 41 0D 0C \r7EA 03 41 0D 0C \r\r>'
...
```

Notice that both the first 'Vehicle Speed' and 'Engine RPM' commands do not use headers.

Then `AT SH 7C0` is isued to enable `2129` custom command.

Subsequently, `2129` and `212A` are executed. Notice that `212A` does not issue a header command because it needs the same header of the previous one.

Before doing again 'Vehicle Speed' and 'Engine RPM', the header is restored to `AT SH 7E0`.

Ircama commented:
Meanwhile I added the feature to restore the default header when closing the connection (this is because if the connection is opened again, it does not set the default header and only sets it when needed, so expects to send commands to ECU 01 until a different header is requested). I rebased the commit.

alistair23 commented on docs/Custom Commands.md:
```diff
@@ -95,7 +96,13 @@ The `ecu` argument is a constant used to filter incoming messages. Some commands
 ### OBDCommand.fast
 
-The `fast` argument tells python-OBD whether it is safe to append a `"01"` to the end of the command. This will instruct the adapter to return the first response it recieves, rather than waiting for more (and eventually reaching a timeout). This can speed up requests significantly, and is enabled for most of python-OBDs internal commands. However, for unusual commands, it is safest to leave this  ...
+The optional `fast` argument tells python-OBD whether it is safe to append a `"01"` to the end of the command. This will instruct the adapter to return the first response it recieves, rather than waiting for more (and eventually reaching a timeout). This can speed up requests significantly, and is enabled for most of python-OBDs internal commands. However, for unusual commands, it is safest to le ...
```
What are you changing on this line?

Ircama replied:
I just added 'optional'. The *optional* `fast` argument...

alistair23 replied:
Ah, that's fine. The patch looks good then, just needs to be tested

alistair23 commented:
Ok, cool. This looks fine to me besides editing the one extra line of documentation.

I'll try and test this today before I merge it.
